아래는 3차원 데이터셋을 만드는 코드입니다.

In [1]:
import numpy as np

np.random.seed(4)
m = 60
w1, w2 = 0.1, 0.3
noise = 0.1

angles = np.random.rand(m) * 3 * np.pi / 2 - 0.5
X = np.empty((m, 3))
X[:, 0] = np.cos(angles) + np.sin(angles)/2 + noise * np.random.randn(m) / 2
X[:, 1] = np.sin(angles) * 0.7 + noise * np.random.randn(m) / 2
X[:, 2] = X[:, 0] * w1 + X[:, 1] * w2 + noise * np.random.randn(m)

In [2]:
from sklearn.decomposition import PCA

아래는 PCA 모델을 사용해서 데이터셋의 차원을 2차원으로 줄이는 코드입니다.<br>
이 때, 사이킷런의 PCA는 데이터를 알아서 중앙에 맞춰주므로<br>
따로 중앙에 맞춰주는 코드는 필요없습니다.

In [3]:
pca = PCA(n_components=2)
X2D = pca.fit_transform(X)

In [4]:
print(X.shape, X2D.shape)

(60, 3) (60, 2)


3차원의 데이터가 2차원으로 축소된 것을 볼 수 있습니다.

PCA 변환기를 데이터셋에 학습시키고 나면 components\_ 속성에 \\(W_{d}\\)의 전치가 담겨있습니다.

In [5]:
pca.components_.T

array([[-0.93636116,  0.34027485],
       [-0.29854881, -0.90119108],
       [-0.18465208, -0.2684542 ]])

explained\_variance\_ratio\_ 변수에는 주성분의 설명된 분산의 비율이 있습니다.<br>
이 비율은 각 주성분의 축을 따라 있는 데이터셋의 분산 비율을 나타냅니다.

In [6]:
pca.explained_variance_ratio_

array([0.84248607, 0.14631839])

전체 데이터셋의 84.2%가 첫 번째 PC를 따라 놓여있고,<br>
14.6%가 두 번째 PC를 따라 놓여있음을 알려줍니다.

세 번째 PC에는 1.2% 미만이 남아있을 것이므로 아주 적은 양의 정보가 있다고 생각해도 됩니다.